Gce cloud provider

mkjelland opened the pull request:
**What this PR does / why we need it**:
Enables GCE cloud provider support:
- Creates GCE service accounts that have the specific permissions needed by the master and worker machines for cloud provider functionality
- Starts the kubelet with cloud provider configuration, this is necessary for the kubelet to attach and detach gce persistent disks
- Adds a default storage class in the master node's startup script
- Instructs users how to deploy a gce ingress controller to their cluster

**Which issue(s) this PR fixes**:
Fixes #81 

**Special notes for your reviewer**:
Once we have a way to deploy addons, the ingress controller and storage class can be deployed in that way.

mkjelland commented:
/assign @krousey

krousey commented:
/ok-to-test

krousey commented on gcp-deployer/machine_setup_configs.yaml:
```diff
@@ -183,6 +208,7 @@ items:
       Environment="KUBELET_NETWORK_ARGS=--network-plugin=kubenet"
       Environment="KUBELET_DNS_ARGS=--cluster-dns=${CLUSTER_DNS_SERVER} --cluster-domain=${CLUSTER_DNS_DOMAIN}"
       EOF
+      sed -i "2iEnvironment=\"KUBELET_EXTRA_ARGS=--cloud-provider=gce --cloud-config=/etc/kubernetes/cloud-config\"" /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
```
Can you add this to the above `cat` command?

krousey commented on gcp-deployer/machine_setup_configs.yaml:
```diff
@@ -70,6 +70,7 @@ items:
       Environment="KUBELET_NETWORK_ARGS=--network-plugin=kubenet"
       Environment="KUBELET_DNS_ARGS=--cluster-dns=${CLUSTER_DNS_SERVER} --cluster-domain=${CLUSTER_DNS_DOMAIN}"
       EOF
+      sed -i "2iEnvironment=\"KUBELET_EXTRA_ARGS=--cloud-provider=gce --cloud-config=/etc/kubernetes/cloud-config\"" /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
```
Add this to the above `cat` command.

krousey commented on gcp-deployer/machine_setup_configs.yaml:
```diff
@@ -112,6 +113,21 @@ items:
+        type: pd-standard
+      EOF
+      kubectl apply --kubeconfig /etc/kubernetes/admin.conf -f /etc/kubernetes/default-storage-class.yaml
+
```
Is this the best place to install this? The API is running at this point, and we should be able to install this from anywhere. We don't want to do this on every master create (in a multi-master scenario). Also, would be nice if we could do this in a place that isn't a bash install script.

It's fine for now, but we will likely want to move away from this.

krousey commented on docs/examples/gce-ingress-controller/ingress-controller.yaml:
```diff
@@ -0,0 +1,165 @@
+---
```
Do you need a document separator here?

krousey commented on docs/examples/gce-ingress-controller/ingress-controller.yaml:
```diff
@@ -0,0 +1,165 @@
+rules:
+- apiGroups: [""]
+  resources: ["secrets", "endpoints", "services", "pods", "nodes", "namespaces", "configmaps", "events"]
+  verbs: ["describe", "get", "list", "watch", "update", "create", "patch"]
```
I'm curious what the ingress controller updates or patches in the above resources. I would thing read-only access would be enough.

mkjelland commented:
/retest

krousey commented on docs/examples/gce-ingress-controller/ingress-controller.yaml:
```diff
@@ -0,0 +1,165 @@
+    k8s-app: glbc
+---
+apiVersion: v1
+kind: ReplicationController
```
Can we change this to `Deployment`? `ReplicationController`s are so Kubernetes v1.1.

krousey commented on docs/examples/gce-ingress-controller/README.md:
```diff
@@ -0,0 +1,12 @@
+
+1. Replace `<YOUR PROJECT ID>` and `<YOUR CLUSTER NAME>` in
+`ingress-controller.yml`.
+1. Run `kubectl create -f ingress-controller.yml`. This will create
```
yaml

krousey commented on pkg/controller/machine/actuator.go:
```diff
@@ -23,6 +23,10 @@ import (
+	// TODO mkjelland move ProvisionClusterDependencies to a cluster actuator
+	// Provision infrastructure that the cluster needs before it
+	// can be created
+	ProvisionClusterDependencies(*clusterv1.Cluster, []*clusterv1.Machine) error
```
Can we not add this here. There's a `machineDeployer` in the gce-deployer package that's more appropriate.

krousey commented on cloud/google/serviceaccount.go:
```diff
@@ -26,68 +23,142 @@ import (
+		"storage.admin",
+		"storage.objectViewer",
+	}
+	WorkerNodeRoles = []string{
```
Why do nodes need these roles? (I'm uninformed)

krousey replied:
If possible, could this be a service account that the ingress controller has instead of every node it could run on? Like the machine controller does:

https://github.com/kubernetes-sigs/cluster-api/blob/master/cloud/google/serviceaccount.go#L75
https://github.com/kubernetes-sigs/cluster-api/blob/master/cloud/google/config/configtemplate.go#L174-L176
https://github.com/kubernetes-sigs/cluster-api/blob/master/cloud/google/config/configtemplate.go#L138-L139
https://github.com/kubernetes-sigs/cluster-api/blob/master/cloud/google/config/configtemplate.go#L144-L146

krousey commented on cloud/google/serviceaccount.go:
```diff
@@ -26,68 +23,142 @@ import (
+	}
 )
 
+// Returns the email address of the service account that should be used
```
Would this be better stored in structured provider status for the cluster? Annotations are fine for now.

krousey replied:
https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/apis/cluster/v1alpha1/cluster_types.go#L99 you can create one for GCE.

mkjelland replied:
I want to create a cluster actuator and add onto the current cluster controller in the near future. Do you think it would be ok to wait until then to create this? I think I'll understand what all belongs in the cluster status better by that point.

krousey commented on cloud/google/serviceaccount.go:
```diff
@@ -26,68 +23,142 @@ import (
+	return nil
+}
+
+func (gce *GCEClient) createServiceAccount(serviceAccountPrefix string, roles []string, cluster *clusterv1.Cluster, initialMachines []*clusterv1.Machine) (string, string, error) {
```
What's the `string, string, error` return values? It's not obvious without digging through the implementation.

krousey commented on cloud/google/serviceaccount.go:
```diff
@@ -26,68 +23,142 @@ import (
+	WorkerNodeServiceAccountPrefix        = "k8s-worker"
+	MachineControllerServiceAccountPrefix = "k8s-machine-controller"
+	MachineControllerSecret               = "machine-controller-credential"
+	ClusterAnnotationPrefix               = "service-account-"
```
We might want to adopt a provider specific prefix. Like `gce.clusterapi.k8s.io/service-account-`.  WDYT?

mkjelland replied:
good idea. Added

mkjelland commented:
@krousey I had to create a post creation step to the machine deployer in order to create a service account for the ingress controller because a key for the service account needs to be stored in a secret for the controller to be able to access it easily. I moved the storage class creation to this step. I also added a step to deploy the ingress controller automatically because I thought it would be weird if the code deployed a service account but not the ingress controller itself. This post creation step will need to be replaced eventually by a different solution for addons, but I like this better for now. Let me know what you think!

krousey commented on cloud/google/machineactuator.go:
```diff
@@ -398,13 +393,41 @@ func (gce *GCEClient) Delete(machine *clusterv1.Machine) error {
+		return fmt.Errorf("error creating service account for ingress controller: %v", err)
+	}
+
+	if len(machines) > 0 {
```
TODO to pull this from the cluster instead

krousey commented:
/lgtm
